KugleFirmware/Libraries/Devices/BatteryBoard/BatteryBoard.cpp

```diff
@@ -0,0 +1,139 @@
+/* Copyright (C) 2018-2019 Thomas Jespersen, TKJ Electronics. All rights reserved.
+ *
+ * This program is free software: you can redistribute it and/or modify  
+ * it under the terms of the GNU General Public License as published by  
+ * the Free Software Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but 
+ * WITHOUT ANY WARRANTY; without even the implied warranty of 
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU 
+ * General Public License for more details. 
+ *
+ * Contact information
+ * ------------------------------------------
+ * Thomas Jespersen, TKJ Electronics
+ * Web      :  http://www.tkjelectronics.dk
+ * e-mail   :  thomasj@tkjelectronics.dk
+ * ------------------------------------------
+ */
+
+#include "BatteryBoard.h"
+#include "SMBus.h"
+#include "Debug.h"
+
+BatteryBoard::BatteryBoard(SMBus::port_t SMBusHardwareResources, uint16_t _chargerCurrentLimit) : chargerCurrentLimit(_chargerCurrentLimit), batteryBoardRamTaskHandle(0), batteryBoardEepromTaskHandle(0)
+{
+
+	// the device can only be accessed when the charger is plugged in!
+	smbus = new SMBus(SMBusHardwareResources, DeviceAddr);
+
+	SetCurrentLimitRAM(chargerCurrentLimit); // set the input limit of the batteryboard to whatever the charger can deliver!
+	//SetCurrentLimitEEPROM(chargerCurrentLimit); // set the input limit of the batteryboard to whatever the charger can deliver!
+
+	// We do this in RAM for 2 reasons!
+	// 1) the EEPROM have limited erase/write cycles
+	// 2) If the change was saved to EEPROM it would only take affect after all sources of power to the battery board is removed (thus it requires physical interaction with the robot, e.g. removing the battery and charger cable)
+	// Therefore a change the the setting
+}
+
+BatteryBoard::~BatteryBoard()
+{
+	if(smbus){delete smbus;}
+	if(batteryBoardRamTaskHandle){vTaskDelete(batteryBoardRamTaskHandle);}
+	if(batteryBoardEepromTaskHandle){vTaskDelete(batteryBoardEepromTaskHandle);}
+}
+
+void BatteryBoard::BatteryBoardRAMThread(void * pvParameters)
+{
+	BatteryBoard * thisObjPtr = (BatteryBoard *)pvParameters;
+
+	while(1){ // loop until the device becomes available and the configuration can be changed
+		if(thisObjPtr->SetCurrentLimitRAM()){ // stop task
+			if (thisObjPtr->batteryBoardRamTaskHandle){ // free memory associated with the task
+				vTaskSuspend(thisObjPtr->batteryBoardRamTaskHandle);
+			}
+		}
+		osDelay(1000);
+	}
+}
+
+void BatteryBoard::BatteryBoardEEPROMThread(void * pvParameters)
+{
+	BatteryBoard * thisObjPtr = (BatteryBoard *)pvParameters;
+
+	while(1){ // loop until the device becomes available and the configuration can be changed
+		if(thisObjPtr->SetCurrentLimitEEPROM()){ // stop task
+			if (thisObjPtr->batteryBoardEepromTaskHandle){ // free memory associated with the task
```

Comment:
Isn't the comment misleading - you are not freeing memory related to the task
But I see how you suspend and resume. However there is almost no overhead in creating and destroying threads. So you could also just destroy the thread completely similar to what is done here: https://github.com/mindThomas/Kugle-Embedded/blob/master/KugleFirmware/Libraries/Applications/BalanceController/BalanceController.cpp#L962